File: app/scripts/research/blind/qwen_mode_probe.mjs
/** Per-call raw-prefix diagnostic through the installed SDK, separate from scoring. */
import {readFile,writeFile,mkdir,stat} from 'node:fs/promises';
import {createRequire} from 'node:module';
import {createHash} from 'node:crypto';
const require=createRequire(import.meta.url);
const {LMStudioClient}=require('/Users/wk/.lmstudio/extensions/plugins/lmstudio/rag-v1/node_modules/@lmstudio/sdk/dist/index.cjs');
const root='/Users/wk/.local/share/means-of-prediction/slides/astra-qwen-nyt-round1-20260912';
const packet=JSON.parse(await readFile(root+'/runtime-mode-sdk-inputs.private.json','utf8'));
const limit=Number(process.argv[2]||1);
const extended=process.argv[3]==='open-free-8192';
const client=new LMStudioClient({baseUrl:'ws://127.0.0.1:1234'});
const model=await client.llm.model(packet.identifier);
const info=await model.getModelInfo();
const exists=async p=>stat(p).then(()=>true,()=>false);
const save=async(p,v)=>writeFile(p,JSON.stringify(v,null,2),{flag:'wx',mode:0o600});
const sha=s=>createHash('sha256').update(s).digest('hex');
for(const row of packet.cases.slice(0,limit)){
  const rendered=await model.applyPromptTemplate(row.messages);
  if(!rendered.endsWith('<think>\n'))throw new Error('Unexpected baseline thinking prefix; refusing guessed rewrite.');
  for(const variant of (extended?['open-free-8192']:['open-grammar','closed-grammar','open-free'])){
    const directory=`${root}/runtime-probes/sdk-mode/${row.caseId}/${variant}`;
    if(await exists(directory+'/result.json'))continue;
    if(await exists(directory+'/started.json'))throw new Error('Uncertain existing probe; no restart.');
    await mkdir(directory,{recursive:true});
    const prompt=variant==='closed-grammar'?rendered.slice(0,-8)+'<think>\n\n</think>\n\n':rendered;
    const config={temperature:0,maxTokens:extended?8192:2048,topKSampling:40,topPSampling:0.95,minPSampling:0.05,repeatPenalty:1.1,contextOverflowPolicy:'stopAtLimit',
      structured:variant.startsWith('open-free')?{type:'none'}:{type:'json',jsonSchema:packet.schema}};
    const tokens=await model.countTokens(prompt);
    if(tokens+config.maxTokens>info.contextLength)throw new Error('Full diagnostic input cannot fit.');
    const request={model:packet.identifier,prompt,config,api:'SDK.complete',baseRenderedSha256:sha(rendered),inputTokens:tokens,modelInfo:info};
    await save(directory+'/request.json',request);
    await save(directory+'/started.json',{at:new Date().toISOString(),pid:process.pid});
    const start=Date.now();
    try{
      const result=await model.complete(prompt,config);
      await save(directory+'/response.json',{content:result.content,reasoningContent:result.reasoningContent,nonReasoningContent:result.nonReasoningContent,
        stats:result.stats,modelInfo:result.modelInfo,loadConfig:result.loadConfig,predictionConfig:result.predictionConfig});
      await save(directory+'/result.json',{status:'response_received',seconds:(Date.now()-start)/1000,variant,inputTokens:tokens,stats:result.stats,
        responseSha256:sha(await readFile(directory+'/response.json')),requestSha256:sha(await readFile(directory+'/request.json'))});
      process.stdout.write(JSON.stringify({caseId:row.caseId,variant,seconds:(Date.now()-start)/1000,stats:result.stats})+'\n');
    }catch(error){
      await save(directory+'/result.json',{status:'failed',exceptionType:error.constructor.name,message:error.message,seconds:(Date.now()-start)/1000,variant});
      process.stdout.write(JSON.stringify({caseId:row.caseId,variant,status:'failed',message:error.message})+'\n');
    }
  }
}

File: app/scripts/research/blind/qwen_mode_probe.py
"""Separate native-API reasoning-mode capability diagnostic; never benchmark data."""
import argparse
import concurrent.futures
import http.client
import json
import os
import time
from pathlib import Path
import qwen_round1 as q


def request_for(item, rule, mode):
    baseline = q.judge_request(rule, q.restore_email(item['email']), (q.ROOT / 'baseline-judge.txt').read_text(), q.load('runtime.json'))
    return {'model': baseline['model'], 'input': baseline['messages'][1]['content'],
            'system_prompt': baseline['messages'][0]['content'], 'reasoning': mode,
            'temperature': 0, 'top_p': 0.95, 'top_k': 40, 'min_p': 0.05,
            'repeat_penalty': 1.1, 'max_output_tokens': 2048, 'store': False,
            'stream': False, 'integrations': []}


def call(request, directory):
    directory.mkdir(parents=True, exist_ok=True)
    if (directory / 'request.json').exists() and q.r.read(directory / 'request.json') != request:
        raise RuntimeError('Probe request changed')
    if (directory / 'result.json').exists(): return q.r.read(directory / 'result.json')
    if (directory / 'started.json').exists(): raise RuntimeError('Uncertain probe; do not restart')
    q.r.save(directory / 'request.json', request)
    q.r.save(directory / 'started.json', {'at': q.r.now(), 'pid': os.getpid()})
    start = time.time(); connection = http.client.HTTPConnection('127.0.0.1', 1234, timeout=2700)
    try:
        connection.request('POST', '/api/v1/chat', json.dumps(request, ensure_ascii=False).encode(), {'Content-Type': 'application/json'})
        response = connection.getresponse(); raw = response.read()
        (directory / 'response.raw').write_bytes(raw)
        data = json.loads(raw); q.r.save(directory / 'response.json', data)
        messages = [x['content'] for x in data.get('output', []) if x.get('type') == 'message']
        try: output = json.loads(''.join(messages)); valid = q.valid_judgment(output)
        except (ValueError, TypeError): output = None; valid = False
        stats = data.get('stats', {})
        result = {'status': 'completed' if response.status == 200 and valid else 'failed', 'output': output,
                  'httpStatus': response.status, 'seconds': time.time()-start,
                  'requestSha256': q.r.hash_value(request), 'rawResponseSha256': q.r.digest(directory / 'response.raw'),
                  'responseSha256': q.r.digest(directory / 'response.json'), 'nativeStats': stats,
                  'reasoningCharacters': sum(len(x.get('content', '')) for x in data.get('output', []) if x.get('type') == 'reasoning'),
                  'returnedModel': data.get('model_instance_id'), 'messageCount': len(messages),
                  'usage': {'prompt_tokens': stats['input_tokens'], 'completion_tokens': stats['total_output_tokens']} if stats else None,
                  'maxTokensPossiblyReached': stats.get('total_output_tokens') == request['max_output_tokens']}
    except Exception as exc:
        result = {'status': 'transport_failed', 'output': None, 'exceptionType': type(exc).__name__,
                  'seconds': time.time()-start, 'requestSha256': q.r.hash_value(request)}
    finally: connection.close()
    q.r.save(directory / 'result.json', result)
    return result


def main(limit, workers):
    os.umask(0o077)
    panel = q.load('runtime-mode-diagnostic-panel.json')['cases'][:limit]
    items = {x['caseId']: x for x in q.load(q.ITEMS)}
    rules = {x['marketId']: x for x in q.load('methods/baseline/rules.json')}
    tasks = [(entry, mode) for entry in panel for mode in ['off', 'on']]
    def one(task):
        entry, mode = task; item = items[entry['caseId']]
        result = call(request_for(item, rules[item['marketId']]['output'], mode), q.ROOT / 'runtime-probes/native-mode' / item['caseId'] / mode)
        row = {**entry, 'mode': mode, **result, 'score': q.score_record(item, {'trial': 1, **result})}
        print(json.dumps({k: row[k] for k in ['caseId', 'mode', 'status', 'httpStatus', 'seconds', 'reasoningCharacters'] if k in row}), flush=True)
        return row
    with concurrent.futures.ThreadPoolExecutor(max_workers=workers) as pool: rows = list(pool.map(one, tasks))
    q.once(f'runtime-mode-diagnostic-results-{limit}.private.json', {'at': q.r.now(), 'rows': rows,
           'qualification': 'Targeted development capability probe. Same frozen rule, whole semantic email and system prompt; native reasoning off/on, no JSON grammar or documented seed option. Endpoint, grammar and seed support differ from the scored baseline. No causal attribution or benchmark promotion from this panel. Native mode is per request; no loaded-instance default changed.'})


if __name__ == '__main__':
    p = argparse.ArgumentParser(); p.add_argument('--limit', type=int, default=1); p.add_argument('--workers', type=int, default=2); a = p.parse_args(); main(a.limit, a.workers)
